myia/info.py

```diff
@@ -0,0 +1,103 @@
+"""Objects and routines to track debug information."""
+
+from typing import Any
+import types
+import threading
+import traceback
+
+
+# We use per-thread storage for the about stack.
+_about = threading.local()
+_about.stack = [None]
+
+
+def current_info():
+    """Return the `DebugInfo` for the current context."""
+    return _about.stack[-1]
+
+
+class DebugInfo(types.SimpleNamespace):
+    """Debug information for an object.
+
+    Attributes:
+        about: DebugInfo of a different object, that this
+            object is derived from in some way
+        relation: How the object relates to `about`
+        save_trace: Whether the trace of the DebugInfo's
+            creation should be saved.
+        trace: The trace at the moment of the DebugInfo's
+            creation, or None.
+
+    """
+
+    def __init__(self, **kwargs):
+        """Construct a DebugInfo object."""
+        self.about = None
+        self.relation = None
+        self.save_trace: bool = False
+        self.trace: Any = None
+        top = current_info()
+        if top:
+            # Only need to look at the top of the stack
+            self.__dict__.update(top.__dict__)
```

Comment:
So you're copying `about` and `relation` as well? Does that mean that two objects can have the same relationship to the same object if `about` and `relation` isn't overwritten elsewhere?

Reply:
Potentially, yes. If a node is replaced by two nodes, you can say that they both have the "replaces" relation with respect to the original. That's coarse information, but better than none.

Reply:
What I meant more is:
```python
with about(about=x, relation='child'):
    with about(about=y):
        n = Constant(2)  # this node is now 'child' of x?
```

Reply:
In this case, no, `relation` will be `None` (the default value for that argument in `about`). Would it be better to require that argument?

`with DebugInfo(about=y)` would have the behavior you describe, though.

Reply:
It's a bit opaque that `DebugInfo` inherits everything from its parent, except if constructed through `about`, in which case the attributes `relation` and `about` are required, or overwritten.

Is it obvious that a child should inherit everything from its parent? For example, using `with DebugInfo(about=y)` copying the relationship of its parent while the `about` field is different, is confusing. Setting `_id` to `None` seems to be another example where this inheriting behavior has to be manually overridden to prevent weird behavior, and in doing so it actually leaks implementation details from the `NamedDebugInfo` subclass.

Is it possible to not inherit everything, but instead just set a `parent` attribute? Or maybe we can have a subclass of `DebugInfo` with a list of attributes that should be inherited (e.g. `location` and `ast`, I imagine).

Reply:
The idea is that if you write `with DebugInfo(...)` you want every `DebugInfo` created in that context to have properties `...`. That's the whole point of doing it. Going through parents would be an annoyance. A whitelist wouldn't work, since you may want to make up new debug properties to help you debug something e.g. `with DebugInfo(my_transform=parameter): my_transform(parameter, graph)`.

A cleaner implementation, though, might be to use a different class than `DebugInfo` as a context manager: then it would never have any fields you don't mean to inherit.

Reply:
That might be a good idea. It seems weird that any assigment to a debug object propagates to children. Only the ones given in the context manager should propagate.

The `about` and `relationship` are an odd one out, but maybe they should just be one field i.e. `about = (x, 'child')`. That way it either fully propagates, or not at all, but you can't have the thing where a relationship suddenly applies to a new object. In that case, maybe you can also get rid of the `about` constructor and just use `DebugInfo` directly?